cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -301,4 +301,19 @@ int accessPathLimit() { result = 5 }
  *
  * This predicate is only used for consistency checks.
  */
-predicate isImmutableOrUnobservable(Node n) { none() }
+predicate isImmutableOrUnobservable(Node n) {
+  // Is the null pointer (or something that's not really a pointer)
+  exists(n.asExpr().getValue())
+  or
+  // Isn't a pointer or is a pointer to const
+  forall(DerivedType dt | dt = n.asExpr().getActualType() |
+    dt.getBaseType().isConst()
+    or
+    dt.getBaseType() instanceof RoutineType
+  )
+  or
+  // Isn't something we can track
+  n.asExpr() instanceof Call
```

Comment:
I'd recommend against this line.  Presumably a `Call` can be a getter, in which case you'd want a postupdatenode to allow a backwards store step.  E.g. in
```
funcWithFlowFromArg0toArg1(source(), objWithFooField.getFoo());
```
you'd want flow to reach `objWithFooField` with an access path of `[.foo]`.

Reply:
Good point. But if these are _consistency_ queries, then I'd expect their results to show me nodes that are unexpectedly missing. A node for `objWithFooField` is missing because the AST data flow isn't feature-complete, but that doesn't mean it's inconsistent.

Reply:
It is the call to `.getFoo()` that may or may not be missing (we're looking at something that's both a call and an argument), and this line suppresses information about such missing postupdatenodes.  If a call in an argument position is missing a postupdatenode and that call has the potential to be a getter, then I think that's definitely an unexpected missing postupdatenode, so I don't think it's a good idea to blanket-filter all calls here.

Reply:
Note that the feature to get this particular flow step working is implemented completely in the shared library as long as sufficient postupdatenodes are present.

Reply:
Thanks for the clarification. I'll try to add the relevant post-update nodes.